REGRESSION(260692@main) 14 Test262 tests fail after this this change

msaboff opened the pull request:
#### 6a323d41f6f0fd052e9daf911fc4b009ad8d1213
<pre>
REGRESSION(260692@main) 14 Test262 tests fail after this this change
<a href="https://bugs.webkit.org/show_bug.cgi?id=253573">https://bugs.webkit.org/show_bug.cgi?id=253573</a>
rdar://106430729

Reviewed by Ross Kirsling.

This fixes three issues introduced in 260692@main:
 1. Improper handling of non-duplicate named groups.  For the case where a non-duplicate
    group appears between two duplicate named groups, the non-duplicate group&apos;s match is
    accessed as a duplicate named group, by using it&apos;s subpatternId to index into the
    duplicate named group indirect area of the output vector.  This could exceed the range
    of the vector, but always provided the wrong result.  This was fixed by restructuring
    the capture group for name processing in both the renamed addCaptureGroupForName() and
    existing setupNamedCaptures().
 2. When parsing a named back reference where the references subpattern may not have been
    closed, we need to do the same analysis we find for numbered back references as found
    in atomBackReference.  That is searching for an unclosed capture parenthesis that
    contains the current back reference.
 3. We shouldn&apos;t try to convert forward references in a lookbehind when an alternative is
    complete as that is premature.  Waiting for the lookbehind assertion is sufficient.

Updated the regexp-duplicate-named-captures.js test, including the group equality checking.

Also renamed the variables *indeces* =&gt; *indices*.

* JSTests/stress/regexp-duplicate-named-captures.js:
(testRegExp):
* Source/JavaScriptCore/runtime/RegExp.cpp:
(JSC::RegExp::finishCreation):
* Source/JavaScriptCore/runtime/RegExp.h:
* Source/JavaScriptCore/yarr/YarrPattern.cpp:
(JSC::Yarr::YarrPatternConstructor::addCaptureGroupForName):
(JSC::Yarr::YarrPatternConstructor::tryConvertingForwardReferencesToBackreferences):
(JSC::Yarr::YarrPatternConstructor::atomParenthesesSubpatternBegin):
(JSC::Yarr::YarrPatternConstructor::atomNamedBackReference):
(JSC::Yarr::YarrPatternConstructor::disjunction):
(JSC::Yarr::YarrPatternConstructor::setupNamedCaptures):
(JSC::Yarr::YarrPattern::compile):
(JSC::Yarr::YarrPattern::YarrPattern):
(JSC::Yarr::YarrPatternConstructor::namedCaptureGroupIdForName): Deleted.
(JSC::Yarr::YarrPatternConstructor::setupDuplicateNamedCaptures): Deleted.
* Source/JavaScriptCore/yarr/YarrPattern.h:
(JSC::Yarr::YarrPattern::resetForReparsing):

Canonical link: <a href="https://commits.webkit.org/261405@main">https://commits.webkit.org/261405@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3164c4bf1254379d498b773e58b35706594ec3dd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111641 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20777 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/251 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/3396 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/120381 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22132 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/11857 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/3144 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117407 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/16440 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/99584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/104503 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/98388 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/88/builds/162 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/104503 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/100141 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13257 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/89/builds/159 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/104503 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/11381 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13747 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/9601 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/101362 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19193 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52155 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/31529 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7942 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15737 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/109401 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26932 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

rkirsling commented on Source/JavaScriptCore/yarr/YarrPattern.cpp:
```diff
@@ -1332,10 +1331,38 @@ class YarrPatternConstructor {
+            PatternAlternative* currentAlternative = m_alternative;
+            ASSERT(currentAlternative);
+
+            while ((currentAlternative = currentAlternative->m_parent->m_parent)) {
```
We've asserted that `currentAlternative` is non-null, but do we need to assert similarly for `currentAlternative ->m_parent`?

msaboff replied:
No, there should always be a parent for an alternative.  For a top level alternative, there is the main disjunction m_body.

rkirsling commented on Source/JavaScriptCore/yarr/YarrPattern.cpp:
```diff
@@ -1803,26 +1827,21 @@ class YarrPatternConstructor {
 
-        m_pattern.m_numDuplicateNamedCaptureGroups = 0;
-
         for (auto& namedGroupIndecies : m_pattern.m_namedGroupToParenIndeces.values()) {
```
nit: Looks like we've got a couple of different misspellings for `indices` in this code—might be nice to clean 'em up right away.

msaboff replied:
I took care of the mispellings.

rkirsling reviewed (approved):
r=me, thanks!

msaboff committed:
REGRESSION(260692@main) 14 Test262 tests fail after this this change
https://bugs.webkit.org/show_bug.cgi?id=253573
rdar://106430729

Reviewed by Ross Kirsling.

This fixes three issues introduced in 260692@main:
 1. Improper handling of non-duplicate named groups.  For the case where a non-duplicate
    group appears between two duplicate named groups, the non-duplicate group's match is
    accessed as a duplicate named group, by using it's subpatternId to index into the
    duplicate named group indirect area of the output vector.  This could exceed the range
    of the vector, but always provided the wrong result.  This was fixed by restructuring
    the capture group for name processing in both the renamed addCaptureGroupForName() and
    existing setupNamedCaptures().
 2. When parsing a named back reference where the references subpattern may not have been
    closed, we need to do the same analysis we find for numbered back references as found
    in atomBackReference.  That is searching for an unclosed capture parenthesis that
    contains the current back reference.
 3. We shouldn't try to convert forward references in a lookbehind when an alternative is
    complete as that is premature.  Waiting for the lookbehind assertion is sufficient.

Updated the regexp-duplicate-named-captures.js test, including the group equality checking.

Also renamed the variables *indeces* => *indices*.

* JSTests/stress/regexp-duplicate-named-captures.js:
(testRegExp):
* Source/JavaScriptCore/runtime/RegExp.cpp:
(JSC::RegExp::finishCreation):
* Source/JavaScriptCore/runtime/RegExp.h:
* Source/JavaScriptCore/yarr/YarrPattern.cpp:
(JSC::Yarr::YarrPatternConstructor::addCaptureGroupForName):
(JSC::Yarr::YarrPatternConstructor::tryConvertingForwardReferencesToBackreferences):
(JSC::Yarr::YarrPatternConstructor::atomParenthesesSubpatternBegin):
(JSC::Yarr::YarrPatternConstructor::atomNamedBackReference):
(JSC::Yarr::YarrPatternConstructor::disjunction):
(JSC::Yarr::YarrPatternConstructor::setupNamedCaptures):
(JSC::Yarr::YarrPattern::compile):
(JSC::Yarr::YarrPattern::YarrPattern):
(JSC::Yarr::YarrPatternConstructor::namedCaptureGroupIdForName): Deleted.
(JSC::Yarr::YarrPatternConstructor::setupDuplicateNamedCaptures): Deleted.
* Source/JavaScriptCore/yarr/YarrPattern.h:
(JSC::Yarr::YarrPattern::resetForReparsing):

Canonical link: https://commits.webkit.org/261405@main
